課題15 model spec作成

aiandrox opened the pull request:
## 概要

<img width="709" alt="スクリーンショット 2020-02-03 18 12 56" src="https://user-images.githubusercontent.com/44717752/73640011-d8942f80-46b0-11ea-9f8d-a8ddc1a5cc61.png">

・model specを追加
　テストデータにFactoryBotを用いる
　letを用いる

## 確認方法

1. スペックを追加したので、`bundle exec rspec spec/models/task_spec.rb --format documentation`を実行してください

## チェックリスト

- [ ] FactoryBotにTaskのデータを追加した
- [ ] モデルスペックを追加した
- [ ] テストが通ることを確認した

## コメント

課題のヒントでは「describeやcontextでケースを分けるほど複雑なテストケースではないので、EverydayRailsと同様にit直後の文章でテストケースを定義してしまいましょう。」とあったのですが、itの中でletを使うと怒られたので、contextを使っています。
もっとスマートな書き方があれば……。

enumで設定されていない値が入力された場合のテストを書こうと思ったのですが、`status: 'not_set'`と入れた時点でエラーが出るようでした。（コメントアウトで書いています。無理なようでしたら削除します）
この項目をテストすることは不可能ということでしょうか？

aiandrox committed:
[Add] gemの追加

aiandrox committed:
[Add] rspecファイル追加

aiandrox committed:
[Fix] rspec-railsのバージョンを3.9に変更

aiandrox committed:
[Add] RSpecでFactroyBotを使えるようにする

aiandrox committed:
[Add] スペックとデータを追加

aiandrox committed:
[Add] モデルスペックを追加

aiandrox committed:
[Fix] 誤字修正

yuji91 commented on spec/models/task_spec.rb:
```diff
@@ -0,0 +1,48 @@
+
+RSpec.describe Task, type: :model do
+  describe 'validation' do
+    let(:new_task) { Task.new(title: title, status: status) }
```
インスタンスを作るだけであれば`FactoryBot.build`が使用できます。
`build(:task, title: title, status: status)`の方が良いかもしれません。

aiandrox replied:
ba4d2cf で修正しました！

yuji91 commented on spec/models/task_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      end
+    end
+
+    it 'タイトルが重複する場合 無効' do
```
`context`作る形で他と統一した方がいいですね

aiandrox replied:
ba4d2cf で修正しました！

yuji91 commented on spec/models/task_spec.rb:
```diff
@@ -0,0 +1,48 @@
+
+    it 'タイトルが重複する場合 無効' do
+      task
+      expect(new_task).to be_invalid
```
`task_with_duplicated_title = build(:task, title: task.title)`と既に作成したタイトルを使ってインスタンスを作成することを明示した方が分かりやすいかなと思います。

aiandrox replied:
ba4d2cf で修正しました！

yuji91 reviewed (changes requested):
コメントしました！

yuji91 commented:
> enumで設定されていない値が入力された場合のテストを書こうと思ったのですが、status: 'not_set'と入れた時点でエラーが出るようでした。（コメントアウトで書いています。無理なようでしたら削除します）
この項目をテストすることは不可能ということでしょうか？

元々 statusはintegerが入るカラムなので、enumと紐づいていないただのStringである文字列の登録は出来ないと思います！

aiandrox committed:
[Fix] 書き換え
・contextを作る形で統一
・build(:task)の書き方に変更
・重複するTaskの名前を:task_with_duplicated_titleに変更

aiandrox committed:
[Add] 設定されていないステータスに関するテストを追加

aiandrox commented:
<img width="709" alt="スクリーンショット 2020-02-03 19 08 41" src="https://user-images.githubusercontent.com/44717752/73644393-a5559e80-46b8-11ea-94f8-58a144c310f1.png">

https://qiita.com/jiz/items/0cd302bb16ce55d231cb
を参考に実装してみました！

yuji91 commented:
`[Fix] 書き換え`だと修正コミットの粒度が大きいと思うので、指摘事項毎に分けたほうが良いかと思います！（次回からで大丈夫です）

yuji91 reviewed (commented):
`let!(:task_with_duplicated_title)`が対象の`context`内のみ影響するか確認をお願いします！

aiandrox commented:
<img width="680" alt="スクリーンショット 2020-02-03 19 38 05" src="https://user-images.githubusercontent.com/44717752/73646414-b99b9a80-46bc-11ea-9435-39c918e6f66c.png">

contextの外では影響しないようです！

itの中に入れてしまうなら、`create(:task)`で書いて、itの外（beforeの代わり）ならletで書くという感じですかね？

yuji91 commented:
> itの中に入れてしまうなら、create(:task)で書いて、itの外（beforeの代わり）ならletで書くという感じですかね？

その認識で大丈夫です！🙆‍♀️

yuji91 commented on spec/models/task_spec.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    context 'タイトルが重複しない場合' do
+      let!(:task_with_duplicated_title) { create(:task) }
```
このletが実行される時はtaskの初回作成時となるので、task_with_duplicated_titleという変数名は正確ではないように思えます。 初回にtaskが作成され、次にtaskを作成する際にtitleが重複しているオブジェクトを格納する変数が`task_with_duplicated_title`であるべきだと思います。

そのため下記の方が自然かなと思います。
```
task = create(:task) # または let!(:task) { create(:task) }
task_with_duplicated_title = build(:task, title: task.title)
expect(task_with_duplicated_title).to be_invalid
```

冗長に感じてしまうようであれば、重複確認のテストケースにbeforeブロックで`let!(:task) {create(:task)}`を記載するのも良いと思います。

aiandrox replied:
3518773 で修正しました！

yuji91 reviewed (changes requested):
コメントしました！

aiandrox committed:
[Fix] taskの名前を変更

aiandrox commented:
<img width="690" alt="スクリーンショット 2020-02-04 18 26 25" src="https://user-images.githubusercontent.com/44717752/73731680-40637c80-477c-11ea-8019-3fb0f6a65e7e.png">

yuji91 commented on spec/factories/tasks.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :task do
+    title { 'タイトル' }
```
`sequence`を使ってFactoryを複数回作成した際にValidationエラーで失敗しないようにしておきたいです。
`create_list`でFactoryを複数作るケースはよく使われるので、それに耐えうる形で`Factory`を定義した方が良いです。

aiandrox replied:
ecdba23 f3cc3ca で修正しました！
この場合、重複をテストするときにわかりづらい気がするので、明示的に`let(:title) { 'title_1' }`と入れた方がわかりやすいでしょうか？

yuji91 reviewed (changes requested):
コメントしました！

aiandrox committed:
[Fix] sequenceを用いてFactoryを作成

aiandrox committed:
[Fix] :titleの変更
最初にcreate(:task)をすると、次のタイトルが変わるため、最初に重複を確かめる

yuji91 commented on spec/factories/tasks.rb:
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  sequence(:title, "title_1")
```
多分この状態だと複数件`create(:task)`を実行した場合にValidationエラーでFactoryの作成が失敗してしまうと思います。
`sequence(:title) { |n| "title-#{n}" }`と、index番号を使って作成するtaskのタイトルが常に一意になるようにしておきましょう。

aiandrox replied:
448552e で修正しました！

aiandrox committed:
[Fix] sequenceの修正

yuji91 reviewed (approved):
LGTM!
